employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/skill/SkillService.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.service.skill;
+
+import java.util.List;
+
+import org.optaweb.employeerostering.domain.Skill;
+import org.optaweb.employeerostering.persistence.SkillRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class SkillService {
+
+    @Autowired
+    private SkillRepository skillRepository;
+
+    public List<Skill> getSkillList(Integer tenantId) {
+        return skillRepository.findAll();
+    }
+
+    public Skill getSkill(Integer tenantId, Long id) {
+        if (skillRepository.findById(id).isPresent())
+            return skillRepository.findById(id).get();
+
+        return new Skill();
+    }
+
+    public void deleteSkill(Integer tenantId, Long id) {
+        if (skillRepository.findById(id).isPresent())
+            skillRepository.deleteById(id);
+    }
+
+    public void createSkill(Integer tenantId, Skill skill) {
+        skillRepository.save(skill);
+    }
+
+    public void updateSkill(Integer tenantId, Skill skill) {
+        if (skill.getId() != null) {
```

Comment:
Soft suggestion: `getId()`'s contract should be defined so that it never returns `null`. Then this check can be removed. If we can't guarantee that it never returns null, it's better to return `Optional<Long>` so that it's harder to avoid the check for presence of the value.

If the `id` is null or empty `Optional` the method should throw `IllegalArgumentException`.

Reply:
From my understanding, all domain objects that inherit from `AbstractPersistable` (including Skill) will have `id = null` until it's persisted in a database. I can't see why `updateSkill()` would be called on a Skill object that's not persisted. I was planning to throw an `EntityNotFoundException` for the null case, but returning `Optional<Long>` may be better. What are your thoughts on this?

Reply:
I would say it depends on whether you consider calling `getId()` on an object that has not yet been persisted (and flushed) an exceptional - or rather illegal - case.

Reply:
@cuijulian Ahh, I see. Good point. So we can't have `final id` at the moment. This is a result of mixing domain model with persistence infrastructure. But this problem exists in the original code so I don't think fixing this is in scope of the migration to Spring Boot. Let's address this later.

If you're interested in seeing how it works when domain model is independent and unaffected by persistence layer, see
- [Location](https://github.com/kiegroup/optaweb-vehicle-routing/blob/master/optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/domain/Location.java) (the domain model class, an entity object that has an identity represented by an unmodifiable id). It's true that it doesn't have an id before it's persisted. And it's because the *entity* doesn't *exist* before being persisted. In other words, the entity object's life cycle starts with being persisted and obtaining a unique id. Before that, everything we have is the user's request to create a new Location entity with the data necessary to create it (an id is not among them).
- [LocationRepository](https://github.com/kiegroup/optaweb-vehicle-routing/blob/master/optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/persistence/LocationRepositoryImpl.java#L44)'s `createLocation()` method. Notice that before the Location is created, we only have the necessary *data* from the client (coordinates and description). Also notice that `LocationEntity` is not *the entity* from domain model perspective (I named it so to honor JPA conventions but I might change that since it's confusing). The `LocationEntity` is a plumbing object that only servers to communicate with the CRUD repository. It's mutable but we don't mind since it never leaks outside the repository API so it doesn't affect the rest of the system, where we mostly work with the domain model where Location is a safe, immutable object that never returns `nulls`.

The vehicle routing code is inspired by [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) ([book](https://www.oreilly.com/library/view/clean-architecture-a/9780134494272/) mildly recommended) and [Domain Driven Design](https://www.oreilly.com/library/view/domain-driven-design-tackling/0321125215/) (recommended).

Reply:
I agree, we should keep the domain model and architecture consistent with original code. `LocationEntity` is an interesting example; we can consider domain model and persistence layer independence at a later step. I'll keep the `id = null` check for now.